### PR TITLE
Add possiblity to tune the index to return at least k nearest neighbors

### PR DESCRIPTION
**Pull request by victor-paltz**
Adding the new parameter `min_nearest_neighbors_to_retrieve`:
> Minimum number of nearest neighbors to retrieve when querying the index. Parameter used only during index hyperparameter finetuning step, it is not taken into account to select the indexing algorithm. This parameter has the priority over the max_index_query_time_ms constraint.

- Add the min_nearest_neighbors_to_retrieve parameter to `build_index `and `tune_index`
- Add new function `get_min_param_value_for_best_neighbors_coverage ` with documentation
- Add tests

### CODE REVIEW

File: autofaiss/external/optimize.py [hunk 1]
```diff
@@ -291,6 +288,118 @@ def get_optimal_quantization(
...
+    nearest   |    /
+    neighbors |   /
+    coverage  |  /
+              | /
+            0 +--[--------------------------]-->  param_value
+                 ^  ^                       ^
+                 |  |                       |
+                 |  min_param_value         |
+                 |                          |
+                 min(parameter_range)      max(parameter_range)
```
Comment (Mbompr) [inline autofaiss/external/optimize.py:315]: 😍

**Review by Mbompr — APPROVED** [review]

File: autofaiss/external/optimize.py [hunk 2]
```diff
@@ -291,6 +288,119 @@ def get_optimal_quantization(
...
+    targeted_nb_neighbors_to_query: int,
+    *,
+    targeted_coverage: float = 0.99,
+    use_gpu: bool = False,
+) -> T:
+    """
+    This function returns the minimal value to set in the index hyperparameters so that,
+    on average, the index retrieves 99% of the requested k=targeted_nb_neighbors_to_query nearest neighbors.
+
+            1 ^      _________________________
```
[the remark this answers is not in the document]
Comment (victor-paltz) [reply autofaiss/external/optimize.py:304]: This is the line of the curve, I could use ----------------- like for the other schema?

File: autofaiss/external/optimize.py [hunk 3]
```diff
@@ -291,6 +288,119 @@ def get_optimal_quantization(
...
+
+    # Intialize the binary search
+    def is_not_meeting_constraint(rank: int) -> bool:
+
+        parameter_value = parameter_range[rank]
+        param_str = hyperparameter_str_from_param(parameter_value)
+        set_search_hyperparameters(index, param_str, use_gpu)
+        nearest_neighbors_coverage = get_nearest_neighbors_coverage(targeted_nb_neighbors_to_query)
+
+        return nearest_neighbors_coverage >= targeted_coverage
```
[the remark this answers is not in the document]
Comment (victor-paltz) [reply autofaiss/external/optimize.py:395]: Good point! The function name is bad
(The test would have caught it otherwise)

File: autofaiss/external/optimize.py [hunk 4]
```diff
@@ -291,6 +288,119 @@ def get_optimal_quantization(
...
+
+    # Find the min param_value that reaches the targeted coverage
+    best_rank = max(0, discrete_binary_search(is_not_meeting_constraint, len(parameter_range)) - 1)
```
[the remark this answers is not in the document]
Comment (hitchhicker) [reply autofaiss/external/optimize.py]: 🤔 When the second argument is -1, it will return lo which is 0.
Anyway this is not very important.

File: autofaiss/external/optimize.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hitchhicker) [reply autofaiss/external/optimize.py:395]: I think that it is not only the problem of function naming actually.
Assume that we change the function to `is_meeting_constraint` and keep the implement unchanged.

That means that when it meets the constraint, we will increase the index by doing `lo = mid + 1`, however, `mid` should the minimal value matching the constraint.

Maybe there is a missing edge case that is not covered by the test.

What do you think ?

[the remark this answers is not in the document]
Comment (hitchhicker) [reply autofaiss/external/optimize.py:395]: Ah, sorry, I have mistaken, it does `lo = mid + 1` when it is ok. No problem then.

**Review by hitchhicker — APPROVED** [review]
LGTM, let's merge it.

**Comment by rom1504** [comment]
cool stuff!
